ledger/byzcoin/mod.go

```diff
@@ -0,0 +1,249 @@
+	}
+}
+
+// GetTransactionFactory implements ledger.Ledger. It ..
+func (ldgr *Ledger) GetTransactionFactory() ledger.TransactionFactory {
+	return ldgr.txFactory
+}
+
+// Listen implements ledger.Ledger. It starts to participate in the blockchain
+// and returns an actor that can send transactions.
+func (ldgr *Ledger) Listen(players mino.Players) (ledger.Actor, error) {
+	bcActor, err := ldgr.bc.Listen(newValidator())
+	if err != nil {
+		return nil, xerrors.Errorf("couldn't start the blockchain: %v", err)
+	}
+
+	err = bcActor.InitChain(&empty.Empty{}, players)
+	if err != nil {
+		return nil, xerrors.Errorf("couldn't initialize the chain: %v", err)
+	}
+
+	err = ldgr.gossiper.Start(players)
+	if err != nil {
+		return nil, xerrors.Errorf("couldn't start the gossiper: %v", err)
+	}
+
+	go ldgr.routine(bcActor, players)
+
+	return newActor(ldgr.gossiper), err
+}
+
+func (ldgr *Ledger) routine(actor blockchain.Actor, players mino.Players) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	blocks := ldgr.bc.Watch(ctx)
+
+	roundTimeout := time.After(initialRoundTime)
+
+	for {
+		select {
+		// TODO: closing
+		case rumor := <-ldgr.gossiper.Rumors():
+			tx, ok := rumor.(Transaction)
+			if ok {
+				ldgr.queue.Add(tx)
+			}
+		case <-roundTimeout:
+			// This timeout has two purposes. The very first use will determine
+			// the round time before the first block is proposed after a boot.
+			// Then it will be used to insure that blocks are still proposed in
+			// of catastrophic failure in the consensus layer (i.e. too many
+			// players offline for a while).
+			go ldgr.proposeBlock(actor, players)
+
+			roundTimeout = time.After(timeoutRoundTime)
+		case block := <-blocks:
+			payload := block.GetPayload().(*BlockPayload)
+
+			txRes := make([]TransactionResult, len(payload.GetTxs()))
+			for i, pb := range payload.GetTxs() {
+				tx, err := ldgr.txFactory.FromProto(pb)
+				if err != nil {
+					return
+				}
+
+				txRes[i] = TransactionResult{
+					txID: tx.(Transaction).hash,
+				}
+			}
+
+			ldgr.queue.Remove(txRes...)
+
+			// This is executed in a different go routine so that the gathering
+			// of transactions can keep on while the block is created.
+			go ldgr.proposeBlock(actor, players)
+
+			roundTimeout = time.After(timeoutRoundTime)
+		}
+	}
+}
+
+func (ldgr *Ledger) proposeBlock(actor blockchain.Actor, players mino.Players) {
+	payload, err := ldgr.makePayload(ldgr.queue.GetAll())
+	if err != nil {
+		fabric.Logger.Err(err).Msg("couldn't make the payload")
+	}
+
+	// Each instance proposes a payload based on the received
+	// transactions but it depends on the blockchain implementation
+	// if it will be accepted.
+	err = actor.Store(payload, players)
+	if err != nil {
+		fabric.Logger.Err(err).Msg("couldn't send the payload")
+	}
+}
+
+func (ldgr *Ledger) makePayload(txs []Transaction) (*BlockPayload, error) {
+	payload := &BlockPayload{
+		Txs: make([]*TransactionProto, 0, len(txs)),
```

Comment:
what is the advantage over using `make([]*TransactionProto, len(txs))` with
```
for i, tx := range txs {
  payload.Txs[I] = ...
}
``` 
?

Reply:
None in that case, the parameter was previously a map and I forgot to update that part ;)